### PR TITLE
Implement List File Share API Support

### LINKED ISSUE #989: List fileshares from AWS
**Issue/Feature Description:**
API to LIST all shares (of the IAM role/tenant) of the AWS
v1/\<tenantId\>/backend/\<backendId\>/shares
**Why this issue to fixed / feature is needed(give scenarios or use cases):**
User should get the list of shares which can be mounted on the target
**How to reproduce, in case of a bug:**

**Other Notes /  Environment Information: (Please give the env information, log link or any useful information for this issue)**

### PR DESCRIPTION
**Pull request by himanshuvar**
**What type of PR is this?**
> /kind new feature

**What this PR does / why we need it**:
This PR is to support the List File Share API. This includes the following:

1. Define FileShare DataModel
2. DB framework Support
3. Add List FileShare API support for routing
4. List File Share for a Backend
5. List all File Shares

**Which issue(s) this PR fixes**:

Fixes #989 #1002 #1003 

**Test Report Added?**:

> /kind TESTED

**Test Report**:

The following testing is done for List File Share API:

- [x] List Backend

- [x] List File Share for Azure Backend

- [x] List File Share for AWS Backend

- [x] List all File Share

```
List Backends:

http://192.168.20.162:8089/v1/94b280022d0c4401bcf3b0ea85870519/backends

{
    "backends": [
        {
            "id": "5ee26198870f6c0001db05b8",
            "tenantId": "94b280022d0c4401bcf3b0ea85870519",
            "userId": "558057c4256545bd8a307c37464003c9",
            "name": "azure-backend",
            "type": "azure-file",
            "region": "Southeast Asia",
            "endpoint": "s3.amazonaws.com",
            "bucketName": "opensds-him"
        },
        {
            "id": "5ee261a4870f6c0001db05b9",
            "tenantId": "94b280022d0c4401bcf3b0ea85870519",
            "userId": "558057c4256545bd8a307c37464003c9",
            "name": "aws-backend",
            "type": "aws-file",
            "region": "ap-south-1",
            "endpoint": "s3.amazonaws.com",
            "bucketName": "opensds-him"
        }
    ],
    "next": 2
}

List File Share for Azure Backend:

http://192.168.20.162:8089/v1/94b280022d0c4401bcf3b0ea85870519/backends/5ee26198870f6c0001db05b8/file/shares

{
    "fileshares": [
        {
            "id": "5ee26313a4fb8b2510259399",
            "createdAt": "2020-06-11T01:12:32.662883Z",
            "updatedAt": "2020-06-11T01:12:34.682329Z",
            "name": "azure-fileshareDemo",
            "description": "Azure FileShare",
            "tenantId": "94b280022d0c4401bcf3b0ea85870519",
            "userId": "558057c4256545bd8a307c37464003c9",
            "backendId": "5ee26198870f6c0001db05b8",
            "backend": "azure-backend",
            "type": "generalPurpose",
            "region": "Southeast Asia",
            "availabilityZone": "Southeast Asia",
            "status": "success"
        }
    ],
    "next": 1
}

List File Share for AWS Backend:

http://192.168.20.162:8089/v1/94b280022d0c4401bcf3b0ea85870519/backends/5ee261a4870f6c0001db05b9/file/shares

{
    "fileshares": [
        {
            "id": "5ee26257a4fb8b2510259397",
            "createdAt": "2020-06-11T01:12:29.662883Z",
            "updatedAt": "2020-06-11T01:12:31.682329Z",
            "name": "aws-fileshareDemo",
            "description": "AWS FileShare",
            "tenantId": "94b280022d0c4401bcf3b0ea85870519",
            "userId": "558057c4256545bd8a307c37464003c9",
            "backendId": "5ee261a4870f6c0001db05b9",
            "backend": "aws-backend",
            "type": "bursting",
            "region": "ap-south-1",
            "availabilityZone": "ap-south-1a",
            "status": "success",
            "tags": [
                {
                    "key": "Name",
                    "value": "FSName"
                },
                {
                    "key": "Department",
                    "value": "Billing"
                }
            ]
        },
        {
            "id": "5ee262d3a4fb8b2510259398",
            "createdAt": "2020-06-11T01:12:29.662883Z",
            "updatedAt": "2020-06-11T01:12:31.682329Z",
            "name": "aws-fileshare-test",
            "description": "AWS FileShare",
            "tenantId": "94b280022d0c4401bcf3b0ea85870519",
            "userId": "558057c4256545bd8a307c37464003c9",
            "backendId": "5ee261a4870f6c0001db05b9",
            "backend": "aws-backend",
            "type": "generalPurpose",
            "region": "ap-south-1",
            "availabilityZone": "ap-south-1b",
            "status": "success",
            "tags": [
                {
                    "key": "Name",
                    "value": "TestFS"
                },
                {
                    "key": "Department",
                    "value": "Testing"
                }
            ]
        }
    ],
    "next": 2
}

List all File Share:

http://192.168.20.162:8089/v1/94b280022d0c4401bcf3b0ea85870519/file/shares

{
    "fileshares": [
        {
            "id": "5ee26257a4fb8b2510259397",
            "createdAt": "2020-06-11T01:12:29.662883Z",
            "updatedAt": "2020-06-11T01:12:31.682329Z",
            "name": "aws-fileshareDemo",
            "description": "AWS FileShare",
            "tenantId": "94b280022d0c4401bcf3b0ea85870519",
            "userId": "558057c4256545bd8a307c37464003c9",
            "backendId": "5ee261a4870f6c0001db05b9",
            "backend": "aws-backend",
            "type": "bursting",
            "region": "ap-south-1",
            "availabilityZone": "ap-south-1a",
            "status": "success",
            "tags": [
                {
                    "key": "Name",
                    "value": "FSName"
                },
                {
                    "key": "Department",
                    "value": "Billing"
                }
            ]
        },
        {
            "id": "5ee262d3a4fb8b2510259398",
            "createdAt": "2020-06-11T01:12:29.662883Z",
            "updatedAt": "2020-06-11T01:12:31.682329Z",
            "name": "aws-fileshare-test",
            "description": "AWS FileShare",
            "tenantId": "94b280022d0c4401bcf3b0ea85870519",
            "userId": "558057c4256545bd8a307c37464003c9",
            "backendId": "5ee261a4870f6c0001db05b9",
            "backend": "aws-backend",
            "type": "generalPurpose",
            "region": "ap-south-1",
            "availabilityZone": "ap-south-1b",
            "status": "success",
            "tags": [
                {
                    "key": "Name",
                    "value": "TestFS"
                },
                {
                    "key": "Department",
                    "value": "Testing"
                }
            ]
        },
        {
            "id": "5ee26313a4fb8b2510259399",
            "createdAt": "2020-06-11T01:12:32.662883Z",
            "updatedAt": "2020-06-11T01:12:34.682329Z",
            "name": "azure-fileshareDemo",
            "description": "Azure FileShare",
            "tenantId": "94b280022d0c4401bcf3b0ea85870519",
            "userId": "558057c4256545bd8a307c37464003c9",
            "backendId": "5ee26198870f6c0001db05b8",
            "backend": "azure-backend",
            "type": "generalPurpose",
            "region": "Southeast Asia",
            "availabilityZone": "Southeast Asia",
            "status": "success"
        }
    ],
    "next": 3
}
```

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by anvithks** [comment]
@himanshuvar In the list Fileshare for Azure backends there is no `tags` key but  in the AWS list there is a `tags` key which is an object array.
I think this should be an empty array in case of Azure fileshares instead of completely removing it from the output.

**Comment by anvithks** [comment]
@himanshuvar @kumarashit Another suggestion:

**Comment by himanshuvar** [comment]
> @himanshuvar @kumarashit Another suggestion:
> The response is currently
> 
> ```
> {
>     "fileshares": [
>         {
>            ...
>          },
>         {
>            ...
>         }
>       ]
> }
> ```
> 
> Our other APIs all return an empty array or an array of items. If possible remove the `fileshares` key.
> So the response will look lilke:
> 
> ```
> [
>         {
>            ...
>          },
>         {
>            ...
>         }
> ]
> ```
> 
> or
> 
> ```
> []
> ```
> 
> in case of empty result.

@anvithks This API is designed/implemented based on the current multi-cloud APIs.
You can check the swagger. https://github.com/sodafoundation/multi-cloud/blob/development/openapi-spec/swagger.yaml

File: api/pkg/file/service.go [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+	s.checkBackendExists(ctx, request, response, backendId)
+
+	filterPathOpts := []string{"backendId"}
+	filterPath, err := common.GetFilterPathParams(request, filterPathOpts)
+	if err != nil {
+		WriteError(response, "Get filter failed: %v\n", http.StatusBadRequest, err)
+		return
+	}
+
+	listfileshareRequest := s.prepareFileShareRequest(request, response)
```
Comment (kumarashit) [inline api/pkg/file/service.go:147]: Better to name this variable as listFileshareRequest

File: file/pkg/model/fileshare.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+	// +optional
+	Encrypted bool `json:"encrypted,omitempty" bson:"encrypted,omitempty"`
+
+	// EncryptionSettings that was used to protect the file share encryption.
+	// +optional
+	EncryptionSettings map[string]string `json:"encryptionSettings,omitempty bson:"encryptionSettings,omitempty"`
+
+	// Metadata should be kept until the semantics between file share and backend storage resource.
+	// +optional
+	Metadata map[string]string `json:"metadata,omitempty bson:"metadata,omitempty"`
```
Comment (kumarashit) [inline file/pkg/model/fileshare.go:98]: What is metadata for?

File: file/pkg/service/service.go [hunk 3]
```diff
@@ -15,14 +15,77 @@
...
+			Protocols:        fs.Protocols,
+			SnapshotId:       fs.SnapshotId,
+			Encrypted:        fs.Encrypted,
+			Metadata:         fs.Metadata,
+		})
+	}
+	out.Fileshares = fileshares
+	out.Next = in.Offset + int32(len(res))
+
+	log.Infof("List File Share successfully, #num=%d, fileshares: %+v\n", len(fileshares), fileshares)
```
Comment (kumarashit) [inline file/pkg/service/service.go:89]:
```suggestion
	log.Infof("Listed fileshare successfully, #num=%d, fileshares: %+v\n", len(fileshares), fileshares)
```

**Comment by kumarashit** [comment]
> @himanshuvar In the list Fileshare for Azure backends there is no `tags` key but in the AWS list there is a `tags` key which is an object array.
> I think this should be an empty array in case of Azure fileshares instead of completely removing it from the output.

Agree. Let's keep it consistent @himanshuvar

**Comment by kumarashit** [comment]
> @himanshuvar @kumarashit Another suggestion:
> The response is currently
> 
> ```
> {
>     "fileshares": [
>         {
>            ...
>          },
>         {
>            ...
>         }
>       ]
> }
> ```
> 
> Our other APIs all return an empty array or an array of items. If possible remove the `fileshares` key.
> So the response will look lilke:
> 
> ```
> [
>         {
>            ...
>          },
>         {
>            ...
>         }
> ]
> ```
> 
> or
> 
> ```
> []
> ```
> 
> in case of empty result.

IMO, the "key" should be kept. This is how the cloud API response also follow:
```
AWS
{
        FileSystems: [{
        CreationTime: 2020-06-03 10:46:16 +0000 UTC,
        CreationToken: "tokenstring",
        Encrypted: false,
        FileSystemId: "fs-4ffe7d9e",
        LifeCycleState: "available",
```
```Azure
<Shares>  
    <Share>  
      <Name>share-name</Name>  
      <Snapshot>Date-Time Value</Snapshot>
      <Properties>  
        <Last-Modified>date/time-value</Last-Modified>  
        <Etag>etag</Etag>  
        <Quota>max-share-size</Quota>
      </Properties>  
      <Metadata>  
        <metadata-name>value</metadata-name>  
      </Metadata>  
    </Share>  
  </Shares>  
```
And also, multcloud APIs follow similar format

File: api/pkg/common/common.go [hunk 4]
```diff
@@ -142,6 +143,19 @@ func GetFilter(request *restful.Request, filterOpts []string) (map[string]string
 	return filter, nil
 }
 
+func GetFilterPathParams(request *restful.Request, filterOpts []string) (map[string]string, error) {
```
Comment (sfzeng) [inline api/pkg/common/common.go:146]: Can we use GetFilter, instead create this new function?

File: api/pkg/file/service.go [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+func (s *APIService) ListFileShareByBackend(request *restful.Request, response *restful.Response) {
+	if !policy.Authorize(request, response, "fileshare:list") {
+		return
+	}
+	log.Info("Received request for File Share List for a Backend.")
+
+	ctx := common.InitCtxWithAuthInfo(request)
+	backendId := request.PathParameter(common.REQUEST_PATH_BACKEND_ID)
+
+	s.checkBackendExists(ctx, request, response, backendId)
```
Comment (sfzeng) [inline api/pkg/file/service.go:138]: What if backend not exist?

File: api/pkg/file/service.go [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+		return
+	}
+	log.Infof("Backend response = [%+v]\n", backendResp)
+}
+
+func (s *APIService) prepareFileShareRequest(request *restful.Request, response *restful.Response) *file.ListFileShareRequest {
+
+	limit, offset, err := common.GetPaginationParam(request)
+	if err != nil {
+		WriteError(response, "Get pagination parameters failed: %v\n", http.StatusInternalServerError, err)
```
Comment (sfzeng) [inline api/pkg/file/service.go:78]: What kind of error code and error message will return from file share service of AWS, AZure, etc. Suggest to list all those error we used, and be compatible with normal way used by those popular cloud vendor.

File: file/cmd/main.go [hunk 7]
```diff
@@ -16,14 +16,22 @@ package main
 
 import (
 	"github.com/micro/go-micro/v2"
+	"os"
```
Comment (sfzeng) [inline file/cmd/main.go:19]: Please format the code.

File: file/pkg/model/fileshare.go [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+	Type string `json:"type,omitempty" bson:"type,omitempty"`
+
+	// The size of the file share requested by the user.
+	// Default unit of file share Size is GB.
+	// +optional
+	Size int64 `json:"size,omitempty" bson:"size,omitempty"`
+
+	// The location that file share belongs to.
+	// +optional
+	Region string `json:"region,omitempty bson:"region,omitempty"`
```
Comment (sfzeng) [inline file/pkg/model/fileshare.go:67]: What's the definition of this Region?

File: file/pkg/model/fileshare.go [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+
+	// The locality that file share belongs to.
+	// +optional
+	AvailabilityZone string `json:"availabilityZone,omitempty bson:"availabilityZone,omitempty"`
```
Comment (sfzeng) [inline file/pkg/model/fileshare.go:71]: What's the definition of this AvailablilityZone?

File: file/pkg/service/service.go [hunk 10]
```diff
@@ -15,14 +15,77 @@
...
-	return &fileService{}
+	return &fileService{
+		fileClient: pb.NewFileService("file", client.DefaultClient),
+	}
+}
+
+func (f *fileService) ListFileShare(ctx context.Context, in *pb.ListFileShareRequest, out *pb.ListFileShareResponse) error {
+	log.Info("Received ListFileShare request.")
+
+	if in.Limit < 0 || in.Offset < 0 {
```
Comment (sfzeng) [inline file/pkg/service/service.go:43]: Better to do this kind of paramater checking in api level.

File: file/pkg/service/service.go (hunk 3 above)
Comment (sfzeng) [inline file/pkg/service/service.go:89]: Suggest to use debug log here, fileshares may be too much to print.

File: api/pkg/file/service.go [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+
+func (s *APIService) ListFileShare(request *restful.Request, response *restful.Response) {
+	if !policy.Authorize(request, response, "fileshare:list") {
+		return
+	}
+	log.Info("Received request for File Share List.")
+
+	ctx := common.InitCtxWithAuthInfo(request)
+
+	listfileshareRequest := s.prepareFileShareRequest(request, response)
```
Comment (sfzeng) [inline api/pkg/file/service.go:112]: Need to check if listfileshareRequest is nil.

File: api/pkg/file/service.go [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+
+func NewAPIService(c client.Client) *APIService {
+	return &APIService{
+		fileClient:   file.NewFileService(fileService, c),
+		backendClient: backend.NewBackendService(backendService, c),
+	}
+}
+
+func UpdateFilter(reqFilter map[string]string, filter map[string]string) error {
+	for k, v := range filter {
```
[the remark this answers is not in the document]
Comment (himanshuvar) [reply api/pkg/file/service.go:49]: Done. 
Thanks

File: api/pkg/file/service.go (hunk 1 above)
Comment (himanshuvar) [reply api/pkg/file/service.go:147]: Done. 
Thanks

File: file/pkg/service/service.go (hunk 3 above)
Comment (himanshuvar) [reply file/pkg/service/service.go:89]: Done. 
Thanks

File: api/pkg/common/common.go (hunk 4 above)
Comment (himanshuvar) [reply api/pkg/common/common.go:146]: GetFilter is for QueryParams. Please check this link.
https://github.com/sodafoundation/multi-cloud/blob/development/api/pkg/common/common.go#L132

For this API, we want to get the PathParam filter for the BackendId

File: api/pkg/file/service.go (hunk 5 above)
Comment (himanshuvar) [reply api/pkg/file/service.go:138]: Code changes Done
Thanks.

File: file/cmd/main.go (hunk 7 above)
Comment (himanshuvar) [reply file/cmd/main.go:19]: Done

File: file/pkg/service/service.go (hunk 10 above)
Comment (himanshuvar) [reply file/pkg/service/service.go:43]: OK Sure.
Code Changes Done
Thanks

File: file/pkg/service/service.go (hunk 3 above)
Comment (himanshuvar) [reply file/pkg/service/service.go:89]: Done
Thanks

File: api/pkg/file/service.go (hunk 11 above)
Comment (himanshuvar) [reply api/pkg/file/service.go:112]: Code Changes Done
Thanks

File: api/pkg/file/service.go (hunk 6 above)
Comment (himanshuvar) [reply api/pkg/file/service.go:84]: Yes,
We will have to provide the uniform API Error Handling Codes for Multi-Cloud APIs.
I have opened an issue for handling errors.
https://github.com/sodafoundation/multi-cloud/issues/1009

File: file/pkg/model/fileshare.go (hunk 2 above)
Comment (himanshuvar) [reply file/pkg/model/fileshare.go:98]: Meta Data is a placeholder for keeping cloud specific parameters like for AWS, Metadata may contain the following:
 NumberOfMountTargets: 3,
 OwnerId: "75027",
 PerformanceMode: "generalPurpose",
 SizeInBytes: {
        Timestamp: 2020-06-03 11:05:32 +0000 UTC,
        Value: 6144,
        ValueInIA: 0,
        ValueInStandard: 6144
  }
FileSystemId: "fs-4ffe7d9e",
LifeCycleState: "available",

File: file/pkg/model/fileshare.go (hunk 9 above)
Comment (himanshuvar) [reply file/pkg/model/fileshare.go:71]: Availability Zone may be required for creating Mount Targets

Do you have any suggestions for Availability Zone?

File: file/pkg/model/fileshare.go (hunk 8 above)
Comment (himanshuvar) [reply file/pkg/model/fileshare.go:67]: Region may be required for creating Mount Targets

Do you have any suggestions for Region?

File: file/pkg/model/fileshare.go (hunk 2 above)
Comment (kumarashit) [reply file/pkg/model/fileshare.go:98]: ok

File: api/pkg/file/service.go (hunk 6 above)
Comment (sfzeng) [reply api/pkg/file/service.go:84]: For object service, we have https://github.com/sodafoundation/multi-cloud/blob/master/s3/error/s3errors.go. File service may have different errors.

File: file/pkg/model/fileshare.go (hunk 8 above)
Comment (sfzeng) [reply file/pkg/model/fileshare.go:67]: Does all cloud file share services have this kind of parameters? I'm just thinking we need to find a normal way. Maybe that depends on the analysis of all parameters for each cloud file share service. If you have such analysis, that would help.

**Comment by kumarashit** [comment]
Is it possible to change to master branch

**Comment by himanshuvar** [comment]
> Is it possible to change to master branch

Yes, The PR is raised to master.
https://github.com/sodafoundation/multi-cloud/pull/1014

**Pull request closed without merging**